### ISSUE TITLE
rootpage_id could not be determined

### ISSUE THREAD

**Issue by papillon6** [dmitryd/typo3-realurl#66]
Hi Dmitry,

RealURL is unable to determine the rootpage_id in a multi-domain-environment, although there are domain-records and defined root-pages. Even if i turn off the automatic configuration and use an individual realurl_conf.php the root-page-ids seem to be ignored (first page of page-tree is used).

I updated from 6.2 to 7.6.4. With 6.2 everything went fine. RealURL-version is 2.0.6.

TYPO3 is installed in a sub-directory, but with proper baseURL and RewriteBase in .htaccess this also worked with TYPO3 6.2.

Are you aware of any bug?

Regards

**Comment by dmitryd** [comment]
Are you sure you have domain record for the domain that you use? If you have a domain record only for `example.com` and try to fetch the page using `www.example.com`, it will not be found because these are different domains.

Also make sure that you use `$GLOBALS['TYPO3_CONF_VARS']` in your config and not `$TYPO3_CONF_VARS`.

**Comment by papillon6** [comment]
Both, www.domain.com and domain.com are defined as domain-records.

Yes, already using $GLOBALS['TYPO3_CONF_VARS'] …

 

Von: Dmitry Dulepov [mailto:notifications@github.com] 
Gesendet: Mittwoch, 9. März 2016 14:04
An: dmitryd/typo3-realurl
Cc: papillon6
Betreff: Re: [typo3-realurl] rootpage_id could not be determined (#66)

 

Are you sure you have domain record for the domain that you use? If you have a domain record only for example.com and try to fetch the page using www.example.com, it will not be found because these are different domains.

Also make sure that you use $GLOBALS['TYPO3_CONF_VARS'] in your config and not $TYPO3_CONF_VARS.

**Comment by dmitryd** [comment]
Do you use `www.domain.com` or `www.domain.com/subdir` in domain records?

**Comment by dmitryd** [comment]
Item number two [here](https://github.com/dmitryd/typo3-realurl/wiki/Notes-for-Integrators#configuring-root-pages-for-realurl)?

**Comment by papillon6** [comment]
Ok, in case someone sutmbles upon this:

I tracked it down. The reason was the sub-directory TYPO3 was installed in. For some reason RealURL fails to fetch the domain-records.

I set up two scenarios:

sub1.domain.de/typo3-installation/
sub2.domain.de/typo3-installation/
--> Does not work - even with "RewriteBase /typo3-installation/" in .htaccess and domain-records (sub1.domain.de and sub2.domain.de) or rootpage_ids set up in realurl_conf.php (no automatic configuration).

sub1.domain.de
sub2.domain.de
--> Works with rootpage_ids set up in realurl_conf.php

$GLOBALS['TYPO3_CONF_VARS']['EXTCONF']['realurl']['_DEFAULT']['pagePath']['rootpage_id'] = ...


Especially when upgrading TYPO3 (without possibility to configure sub-domains or new domains for the development-installation) an installation in a sub-directory was helpful and working (at least with 4.5 and 6.2).

@Dmitry:
Thanks for your help!

**Comment by dmitryd** [comment]
Domain records are in the database, they are not affected by subdirectolries. However if you had wrong domain records (like `www.domain.de/subdir/` instead of `www.domain.de`), realurl will not be able to find the domain because domain is just the domain name and does not include any subdirectories. This is described in the wiki.

**Comment by papillon6** [comment]
As mentioned i was using sub1.domain.de and sub2.domain.de as domain-records (without sub-folder). Only TYPO3 was installed in a sub-directory ("/typo3-installation/" in my example). The configuration was set up correclty I think.

In case you have a test-environment with multiple domains at hand, try to move TYPO3 into a sub-folder to see whether RealUrl works. (I did not dig into the source-code yet.)

**Comment by dmitryd** [comment]
I never used TYPO3 inside a subdir. It does not sound right to me to use an enterprise CMS in a subdirectory. I am always proud that it is my only CMS for each site :smile:

Realurl does not do anything about files except for the inclusion of config. Config is included using standard TYPO3 method, so if TYPO3 works in the subdir, config inclusion should also work. Here is the code:

```
			$realurlConfigurationFile = trim($configuration['configFile']);
			if ($realurlConfigurationFile && @file_exists(PATH_site . $realurlConfigurationFile)) {
				/** @noinspection PhpIncludeInspection */
				require_once(PATH_site . $realurlConfigurationFile);
			}
```

To check if your config is included correctly, you can use `System>Configuration` module. Could you check if it is alright there?

**Comment by tieupmedia** [comment]
Although this topic is three years old, I was actually in the same situation. After searching for a while with absolut no reasonable result, I was glad to see the problem documented here :)

After some further testing, it became clear to me, that this problem isn't related to realurl. I figured out that, if you add an valid page url, e.g. _www.example.com/subdirectory/imprint_, the correct page tree is choosen. But even after uninstalling realurl the entrypoint didn't lead to the right page tree.

So I searched for the way, the core is involved and this took me a great step forward:

In the function _findDomainRecord() of the TYPO3\CMS\Frontend\Controller\TypoScriptFrontendController_ the function _getDomainStartPage of the TYPO3\CMS\Frontend\Page\PageRepository_ is called (line 1895 [or 1887]).

As second parameter _\TYPO3\CMS\Core\Utility\GeneralUtility::getIndpEnv('SCRIPT_NAME')_ is passed. In the webservers document root this is naturally **/index.php** If TYPO3 resides in a subdirectory, it's **/subdirectory/index.php**

Inside of getDomainStartPage the domain to query the rootpage id is resolved. This is build with the given http host plus the resulting path, after stripping away the part from the last slash to the end of the script name (line 870). In root folders the remaining path is '', so it's just **www.example.com** In subdirectories the result is **www.example.com/subdirectory**

If you create a domain record with exactly www.example.com/subdirectory - with no slash at the end - it works in a subdirectory.

But I didn't get the point, what the difference to version < 6.x can be.
